docs/tutorial_custom_controller.md

```diff
@@ -0,0 +1,151 @@
+# Custom Controller Tutorial
+
+Mobly enables users to control custom hardware devices (e.g., smart lights, switches) by creating custom controller modules. This tutorial explains how to implement a production-ready custom controller.
+
+## Controller Module Interface
+
+A Mobly controller module must implement the following top-level functions:
+
+* **`create(configs)`**: Instantiates controller objects from the configuration.
+* **`destroy(objects)`**: Cleans up resources when the test ends.
+* **`get_info(objects)`**: Returns device information for the test report.
+
+## Implementation Example
+
+The following example demonstrates a custom controller for a **Smart Light**, featuring type hinting, input validation, and fault-tolerant cleanup.
+
+### 1. Controller Module (`smart_light.py`)
+
+Save this code as `smart_light.py`.
+
+```python
+"""Mobly controller module for a Smart Light."""
+
+import logging
+from typing import Any, Dict, List
+
+# The key used in the config file to identify this controller.
+MOBLY_CONTROLLER_CONFIG_NAME = "SmartLight"
+
+class SmartLight:
+    """A class representing a smart light device.
+    
+    Attributes:
+        name: String, the name of the device.
```

Comment:
Omit the type in the description if it's already in the type hint to avoid redundancy.